src/engagement_db_to_analysis/engagement_db_to_analysis.py

```diff
@@ -156,6 +160,262 @@ def _fold_messages_by_uid(user, messages_traced_data):
+            messages_by_column[message.participant_uuid] = []
+        messages_by_column[message.participant_uuid].append(column_td)
+
+    # Pass 2: Update each converted rqa message with the demographic messages
+    for msg_td in messages_traced_data:
+        # Skip this message if it's not a demographic.
+        message = Message.from_dict(dict(msg_td))
+        analysis_dataset_config = _analysis_dataset_config_for_message(analysis_config.dataset_configurations, message)
+        if analysis_dataset_config.dataset_type != DatasetTypes.DEMOGRAPHIC:
+            continue
+
+        # Add this demographic to each of the column-view rqa message TracedData for this participant.
+        # (Use messages_by_column.get() because we might have demographics for people who never sent an RQA message).
+        for column_td in messages_by_column.get(message.participant_uuid, []):
+            _add_message_to_column_td(user, msg_td, column_td, analysis_config)
+
+    flattened_messages = []
+    for msgs in messages_by_column.values():
+        flattened_messages.extend(msgs)
+    return flattened_messages
+
+
+def _convert_to_participants_column_format(user, messages_traced_data, analysis_config):
+    """
+    Converts a list of messages traced data into "column-view" format by participant.
+
+    Returns one "column-view" TracedData object for each participant, with all rqa and demographic messages from the
+    same participant appended to each TracedData.
+
+    :param user: Identifier of user running the pipeline.
+    :type user: str
+    :param messages_traced_data: Messages traced data to convert.
+    :type messages_traced_data: list of core_data_modules.traced_data.TracedData
+    :param analysis_config: Configuration for the conversion.
+    :type analysis_config: src.engagement_db_to_analysis.configuration.AnalysisConfiguration
+    :return: Messages organised by participant into column-view format  suitable for further analysis.
+    :rtype: list of core_data_modules.traced_data.TracedData
+    """
+    participants_by_column = dict()  # of participant_uuid -> participant traced data in column view
+    for msg_td in messages_traced_data:
+        message = Message.from_dict(dict(msg_td))
+
+        # If we've not seen this participant before, create an empty Traced Data to represent them.
+        if message.participant_uuid not in participants_by_column:
+            participants_by_column[message.participant_uuid] = TracedData(
+                {"participant_uuid": message.participant_uuid},
+                Metadata(user, Metadata.get_call_location(), TimeUtils.utc_now_as_iso_string())
+            )
+
+        # Add this message to the relevant participant's column-view TracedData.
+        participant = participants_by_column[message.participant_uuid]
+        _add_message_to_column_td(user, msg_td, participant, analysis_config)
+
+    return participants_by_column.values()
```

Comment:
for consistency, lets return a normal list of participants td instead of values() view object list.

Reply:
thanks, forgot py 3 created views